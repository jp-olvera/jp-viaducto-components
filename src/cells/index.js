export * from "./Avatar";
export * from "./Button";
export * from "./Spacer";
