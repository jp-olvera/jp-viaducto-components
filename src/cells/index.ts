export * from './Anchor';
export * from './Avatar';
export * from './Button';
export * from './BareButton';
export * from './CheckBox';
export * from './Container';
export * from './DatePicker';
export * from './Dropdown';
export * from './Grid';
export * from './FormItem';
export * from './Hideable';
export * from './Input';
export * from './Overlay';
export * from './Paragraph';
export * from './Pill';
export * from './Progress';
export * from './Radio';
export * from './Range';
export * from './ResponsivePadding';
export * from './SelectItem';
export * from './Spacer';
export * from './State';
export * from './Switch';
export * from './Tab';
export * from './Table';
export * from './Textarea';
export * from './TimePicker';
export * from './Title';
export * from './Tooltip';
export * from './Keyboard';
