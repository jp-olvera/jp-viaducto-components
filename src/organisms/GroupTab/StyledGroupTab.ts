import styled from 'styled-components';
import StyledTab from '../../cells/Tab/StyledTab';

export const StyledGroupTab = styled.div < any > `
  * {
    box-sizing: border-box;
  }
  display: flex;
  flex-direction: column;
  & ${StyledTab} {
    display: flex;
    justify-content: center;
    align-items: center;
    .tab-text {
      overflow: auto;
      text-align: center !important;
      display: flex;
      justify-content: center;
      align-items: center;
      & > p {
        color: ${(p) => p.configuration.colors[p.tabType].hover} !important;
      }
    }
    &:after {
      width: 0 !important;
      height: 0 !important;
    }
  }
  & .tabs {
    display: flex;
    gap: ${(p) => p.configuration.spacing[p.spacing]};
  }
  & .line {
    transition: 0.2s
      ${(p) => p.transition || p.configuration.transitionTimingFunction};
    height: 0.188rem;
    background-color: ${(p) => p.configuration.colors[p.tabType].hover} !important;
    position: sticky;
    bottom: ${(p) => (p.position === 'top' ? '100%' : 0)};
    left: 0;
  }
`;
