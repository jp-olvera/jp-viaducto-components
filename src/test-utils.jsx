import React from 'react';
import { render } from '@testing-library/react';
import { ConfigProvider } from './providers';

const AllTheProviders = ({ children }) => {
  return <ConfigProvider>{children}</ConfigProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
