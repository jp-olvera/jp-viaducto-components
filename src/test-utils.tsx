import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ConfigProvider } from './providers';

const AllTheProviders = ({ children }: any) => (
  <ConfigProvider>{children}</ConfigProvider>
);

jest.mock('./cells/Input/amex.svg', () => null);
jest.mock('./cells/Input/mastercard.svg', () => null);
jest.mock('./cells/Input/visa.svg', () => null);
jest.mock('./cells/Dropdown/sorting.svg', () => null);
jest.mock('./cells/BareButton/close.svg', () => null);

const customRender = (ui, options?) => render(ui, { wrapper: AllTheProviders, ...options });
// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
