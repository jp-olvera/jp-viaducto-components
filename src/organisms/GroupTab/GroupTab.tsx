import React, { useContext, useState } from 'react';
import { StyledGroupTab } from './StyledGroupTab';

import { ConfigContext } from '../../providers';
/**
 *
 * @param {Tab} children Tab component
 * @param {string} type Set the color variant (type definition colors) for the tabs
 * @param {string} horizontalSpacing Set the horizontal spacing taking the tab content as reference
 * @param {string} verticalSpacing Set the horizontal spacing taking the tab content as reference
 * @param {string} fontSize Set the font size
 * @param {string} position Set the line in position selected
 * @param {string} transition Overrides the transitionTimingFunction
 */

interface GTI {
  children: any;
  type?: string;
  horizontalSpacing?: string;
  verticalSpacing?: string;
  fontSize?: string;
  position?: string;
  transition?: string;
}

const GroupTab = ({
  children,
  type = 'tab',
  horizontalSpacing = 'md',
  verticalSpacing = 'sm',
  fontSize = 'md',
  position = 'bottom',
  transition,
  ...rest
}: GTI) => {
  const { configuration } = useContext(ConfigContext);
  const [getPosition, setPosition] = useState(0);
  const nodes = React.Children.toArray(children).filter((child) => React.isValidElement(child));
  const translate = (index: number) => {
    setPosition(index * 100);
  };
  const pos = position === 'top' ? 'top' : 'bottom';
  return (
    <StyledGroupTab
      horizontalSpacing={horizontalSpacing}
      verticalSpacing={verticalSpacing}
      configuration={configuration}
      count={nodes.length}
      transition={transition}
      type={type}
      position={position}
      {...rest}
    >
      <div className='tabs'>
        {React.Children.toArray(children).map((child: any, index: number) => React.cloneElement(child, {
          horizontalSpacing,
          verticalSpacing,
          lineWidth: '0',
          fontSize,
          position: pos,
          type,
          index,
          active: index === getPosition / 100,
          onClick: () => translate(index),
        }))}
      </div>
      <div
        className='line'
        style={{ transform: `translateX(${getPosition}%)` }}
      />
    </StyledGroupTab>
  );
};

export default GroupTab;
