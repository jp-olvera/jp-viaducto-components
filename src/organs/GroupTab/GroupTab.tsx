import React, {
  useContext, useEffect, useRef, useState,
} from 'react';
import { ConfigProps } from 'ballena-types';
import { StyledGroupTab } from './StyledGroupTab';

import { ConfigContext } from '../../providers';

/** GroupTab component */
interface GTI {
  /** Tab component */
  children: any;
  /** Set the color variant (type definition colors) for the tabs */
  tabType?:
    | 'primary'
    | 'secondary'
    | 'info'
    | 'success'
    | 'warning'
    | 'danger'
    | 'tab';
  /** Set the horizontal spacing taking the tab content as reference */
  horizontalSpacing?:
    | 'none'
    | 'nano'
    | 'micro'
    | 'tiny'
    | 'xs'
    | 'sm'
    | 'md'
    | 'lg'
    | 'xl'
    | 'xxl'
    | 'xxxl';
  /** Set the horizontal spacing taking the tab content as reference */
  verticalSpacing?:
    | 'none'
    | 'nano'
    | 'micro'
    | 'tiny'
    | 'xs'
    | 'sm'
    | 'md'
    | 'lg'
    | 'xl'
    | 'xxl'
    | 'xxxl';
  /** Set the spacing between tabs */
  spacing?:
    | 'none'
    | 'nano'
    | 'micro'
    | 'tiny'
    | 'xs'
    | 'sm'
    | 'md'
    | 'lg'
    | 'xl'
    | 'xxl'
    | 'xxxl';
  /** Set the font size */
  fontSize?: 'xxs' | 'xs' | 'sm' | 'md' | 'lg';
  /** Set the line in position selected */
  position?: 'bottom' | 'top';
  /** Overrides the transitionTimingFunction */
  transition?: string;
  /** Triggers a function changing the tab */
  onTabChange?: Function;
  /** REM base to match any change, base 16 */
  base?: number;
}

/**
 * GroupTab component
 * @param {Tab} children Tab component
 * @param {string} tabType Set the color variant (type definition colors) for the tabs
 * @param {string} horizontalSpacing Set the horizontal spacing taking the tab content as reference
 * @param {string} verticalSpacing Set the horizontal spacing taking the tab content as reference
 * @param {string} fontSize Set the font size
 * @param {string} position Set the line in position selected
 * @param {string} transition Overrides the transitionTimingFunction
 * @param {string} onTabChange Triggers a function changing the tab
 * @param {number} base  REM base to match any change, base 16
 */

const GroupTab = ({
  children,
  tabType = 'tab',
  horizontalSpacing = 'md',
  verticalSpacing = 'sm',
  fontSize = 'md',
  position = 'bottom',
  transition = 'ease',
  spacing = 'none',
  onTabChange,
  base = 16,
  ...rest
}: GTI & React.HTMLAttributes<HTMLDivElement>) => {
  const { configuration } = useContext(ConfigContext);
  const [getPosition, setPosition] = useState<number>(0);
  const [getWidth, setWidth] = useState<number>(0);
  const [place, setPlace] = useState<number>(0);
  const nodes = React.Children.toArray(children).filter((child) => React.isValidElement(child));
  const ref = useRef<HTMLDivElement>(null);
  const translate = (index: number) => {
    setPosition(index * 100);
  };
  const pos = position === 'top' ? 'top' : 'bottom';
  const space: string = spacing || 'none';
  useEffect(() => {
    onload(
      ref,
      getPosition / 100,
      configuration,
      space,
      setPlace,
      setWidth,
      base,
    );
  }, [getPosition, space, horizontalSpacing]);
  const toPx: number = parseFloat(
    configuration.spacing[space].match(/[-]{0,1}[\d]*[.]{0,1}[\d]+/g)[0],
  ) * base;

  return (
    <StyledGroupTab
      horizontalSpacing={horizontalSpacing}
      verticalSpacing={verticalSpacing}
      configuration={configuration}
      count={nodes.length}
      transition={transition}
      tabType={tabType}
      position={position}
      spacing={spacing || 'none'}
      index={getPosition / 100}
      ref={ref}
      {...rest}
    >
      <div className='tabs'>
        {React.Children.toArray(children).map((child: any, index: number) => React.cloneElement(child, {
          horizontalSpacing,
          verticalSpacing,
          lineWidth: '0',
          fontSize,
          position: pos,
          tabType,
          index,
          active: index === getPosition / 100,
          onClick: (e: any) => {
            /* istanbul ignore else */
            if (onTabChange) onTabChange(e);
            translate(index);
          },
        }))}
      </div>
      <div
        className='line'
        style={{
          marginLeft: getPosition === 0 ? 0 : `calc(${toPx}px + ${place}px)`,
          width: `${getWidth}px`,
        }}
      />
    </StyledGroupTab>
  );
};

export default GroupTab;

export const onload = (
  ref: React.RefObject<HTMLDivElement>,
  getPosition: number,
  configuration: ConfigProps,
  spacing: string,
  setPlace: Function,
  setWidth: Function,
  base: number = 16,
) => {
  if (ref && ref.current) {
    if (getPosition === 0) {
      setPlace(0);
    } else {
      let counter: number = 0;
      for (let i = 0; i < getPosition; i++) {
        counter += ref.current.children[0].children[i].clientWidth;
        if (i > 0) {
          counter
            += parseFloat(
              configuration.spacing[spacing].match(
                /[-]{0,1}[\d]*[.]{0,1}[\d]+/g,
              )[0],
            ) * base;
        }
      }
      setPlace(counter);
    }
    setWidth(ref.current.children[0].children[getPosition].clientWidth);
  }
};
