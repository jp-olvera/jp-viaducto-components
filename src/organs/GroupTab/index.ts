import GroupTab from './GroupTab';

export { GroupTab };
