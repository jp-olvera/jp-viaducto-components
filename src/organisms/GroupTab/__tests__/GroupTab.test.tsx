/* eslint-env jest */

import React from 'react';
import { render, fireEvent } from '../../../test-utils';
import { GroupTab } from '..';
import { onload } from '../GroupTab';
import { Tab } from '../../../cells';

describe('<GroupTab/>', () => {
  const Tabs = () => (
    <>
      <Tab text='Tab' />
      <Tab text='Testing' />
      <Tab text='Love' />
    </>
  );
  test('should render properly', () => {
    const { container } = render(
      <GroupTab>
        <Tabs />
      </GroupTab>,
    );
    expect(container).toBeVisible();
  });
  test('should render with top line', () => {
    const { container } = render(
      <GroupTab position='top'>
        <Tabs />
      </GroupTab>,
    );
    expect(container).toBeVisible();
  });
  test('should change tab', () => {
    const { container, queryByText } = render(
      <GroupTab position='top' spacing='lg'>
        <Tab text='Zombie' />
        <Tab text='Act' />
        <Tab text='Dinner' />
      </GroupTab>,
    );
    const tab = queryByText('Dinner');
    fireEvent.click(tab);
    fireEvent.click(queryByText('Zombie'));
    const line: HTMLDivElement = container.querySelector('.line');
    expect(line).toHaveStyle('marginLeft: 0rem');
  });
  describe('onload function', () => {
    const A = () => {
      const ref = React.useRef<HTMLDivElement>(null);
      return <div ref={ref}>aaaa</div>;
    };
    test('should call function', () => {
      expect(
        onload(
          A.bind({}),
          0,
          'lg',
          { spacing: { lg: '1rem' } },
          () => {},
          () => {},
        ),
      ).not.toBeNaN();
    });
    test('should call function with none spacing', () => {
      expect(
        onload(
          A.bind({}),
          0,
          'none',
          { spacing: { lg: '1rem', none: '0rem' } },
          () => {},
          () => {},
        ),
      ).not.toBeNaN();
    });
    test('should call function with undefined spacing', () => {
      expect(
        onload(
          A.bind({}),
          0,
          'aaaa',
          { spacing: { lg: '1rem', none: '0rem' } },
          () => {},
          () => {},
        ),
      ).not.toBeNaN();
    });
  });
});
