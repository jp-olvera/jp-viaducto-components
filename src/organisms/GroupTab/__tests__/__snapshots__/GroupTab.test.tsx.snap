// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupTab/> should change tab 1`] = `
<p
  class="StyledParagraph-sc-dbd4lw-0 dxtxsJ"
  color="#01579B"
>
  Dinner
</p>
`;
